kernel/src/scan/mod.rs

```diff
@@ -70,24 +70,17 @@ impl ScanBuilder {
         }
     }
 
-    /// Predicates specified in this crate's [`Expression`] type.
+    /// Optionally provide an [`Expression`] to filter rows. For example, using the predicate `x <
+    /// 4` to return a subset of the rows in the scan which satisfy the filter. If `predicate_opt`
+    /// is `None`, this is a no-op.
     ///
-    /// Can be used to filter the rows in a scan. For example, using the predicate
-    /// `x < 4` to return a subset of the rows in the scan which satisfy the filter.
-    pub fn with_predicate(mut self, predicate: Expression) -> Self {
-        self.predicate = Some(predicate);
+    /// NOTE: The filtering is best-effort and can produce false positives (rows that should should
+    /// have been filtered out but were kept).
+    pub fn with_predicate(mut self, predicate: impl Into<Option<Expression>>) -> Self {
```

Comment:
question: how did you decide on `Expression` vs `Arc<Expression>` as method parameter?

Reply:
No strong reason, just convenience.

This method turns out to be a main entry point from the outside world. Mostly unit tests and examples, which anyway create the expressions from scratch. Including FFI where we _have_ to create a new instance because native engines can't instantiate an Arc even if they wanted to.

Do you think there's a non-trivial risk that connector writers might get frustrated because they have to repeatedly deep-clone an `ExpressionRef` they already had lying around for some reason? Maybe dashboard queries or something?

Reply:
hm I don't have enough intuition/information to answer that - seems fine for now given that it's our preference to pass an owned `Expression` (from usability in UT etc perspective) but definitely wanted to highlight the possible risk you just mentioned: that the engine has an `Arc<Expression>` laying around and wants to hand it out instead of forcing an owned Expression be passed into this method.

TLDR it gives _us_ (kernel) the flexibility to share the expression after this point but not above it (engine sharing it). right? perhaps we just ship this and change if anyone comes asking :) just noticed this and wanted to ask!

Reply:
Yeah, seems this could be `impl Into<Option<ExpressionRef>>`?